tests/SpecTests/ClientSideEncryptionSpecTest.php

```diff
@@ -569,76 +557,173 @@ public function testCorpus($schemaMap = true)
-        $this->assertSame('test', $clientEncryption->decrypt($encrypted));
-
-        // Test 3
-        $keyId = $clientEncryption->createDataKey('aws', ['masterKey' => $masterKeyConfig + ['endpoint' => 'kms.us-east-1.amazonaws.com']]);
-        $encrypted = $clientEncryption->encrypt('test', ['algorithm' => ClientEncryption::AEAD_AES_256_CBC_HMAC_SHA_512_DETERMINISTIC, 'keyId' => $keyId]);
-        $this->assertSame('test', $clientEncryption->decrypt($encrypted));
+        ]);
 
-        // Test 4
-        $keyId = $clientEncryption->createDataKey('aws', ['masterKey' => $masterKeyConfig + [ 'endpoint' => 'kms.us-east-1.amazonaws.com:443']]);
-        $encrypted = $clientEncryption->encrypt('test', ['algorithm' => ClientEncryption::AEAD_AES_256_CBC_HMAC_SHA_512_DETERMINISTIC, 'keyId' => $keyId]);
-        $this->assertSame('test', $clientEncryption->decrypt($encrypted));
+        $clientEncryptionInvalid = $client->createClientEncryption([
+            'keyVaultNamespace' => 'keyvault.datakeys',
+            'kmsProviders' => [
+                'azure' => Context::getAzureCredentials() + ['identityPlatformEndpoint' => 'example.com:443'],
+                'gcp' => Context::getGCPCredentials() + ['endpoint' => 'example.com:443'],
+            ],
+        ]);
 
-        // Test 5
-        try {
-            $clientEncryption->createDataKey('aws', ['masterKey' => $masterKeyConfig + [ 'endpoint' => 'kms.us-east-1.amazonaws.com:12345']]);
-            $this->fail('Expected exception to be thrown');
-        } catch (ConnectionException $e) {
-        }
+        $test($this, $clientEncryption, $clientEncryptionInvalid);
+    }
 
-        // Test 6
-        try {
-            $clientEncryption->createDataKey('aws', ['masterKey' => $masterKeyConfig + [ 'endpoint' => 'kms.us-east-2.amazonaws.com']]);
-            $this->fail('Expected exception to be thrown');
-        } catch (RuntimeException $e) {
-            $this->assertStringContainsString('us-east-1', $e->getMessage());
-        }
+    public static function customEndpointProvider()
+    {
+        $awsMasterKey = ['region' => 'us-east-1', 'key' => 'arn:aws:kms:us-east-1:579766882180:key/89fcc2c4-08b0-4bd9-9f25-e30687b580d0'];
+        $azureMasterKey = ['keyVaultEndpoint' => 'key-vault-csfle.vault.azure.net', 'keyName' => 'key-name-csfle'];
+        $gcpMasterKey = [
+            'projectId' => 'devprod-drivers',
+            'location' => 'global',
+            'keyRing' => 'key-ring-csfle',
+            'keyName' => 'key-name-csfle',
+            'endpoint' => 'cloudkms.googleapis.com:443',
+        ];
 
-        // Test 7
-        try {
-            $clientEncryption->createDataKey('aws', ['masterKey' => $masterKeyConfig + [ 'endpoint' => 'example.com']]);
-            $this->fail('Expected exception to be thrown');
-        } catch (RuntimeException $e) {
-            $this->assertStringContainsString('parse error', $e->getMessage());
-        }
+        return [
```

Comment:
I noticed in `provideBSONSizeLimitsAndBatchSplittingTests` you were using `yield ...` instead of a large array, which removed a level of nesting. Would it make sense to do that here?

Also, in `provideBSONSizeLimitsAndBatchSplittingTests` you didn't yield a key but did have comments for the test number from the spec. I suppose that could be changed to: `yield 'Test 1' => static function...` to take advantage of named data sets.

Reply:
Good point. Changed both methods to consistently use `yield 'Test ...' => ...`